### PR TITLE
Clean up incorrect article flags

### LINKED ISSUE #436: Non letter characters can be added to an article's tags
**Newsbeuter version **: 2.10

Observed in PR #420, it is possible to flag an article with "2" or "£" by including them in the flag list passed to OP_EDITFLAGS, which is inconsistent with the documentation limiting tags to letters.

### PR DESCRIPTION
**Pull request by tkerdonc**
This modifies the rss_item::sort_flags function in rss.cpp in order to
correct two bugs in the way it cleans up the input flags. Changes are :
- sort_flags no longer fails to remove more than one following or identiqual
  non alphabetic characters.
- sort_flags no longer removes extra characters when a letter appears
  twice in the flag list.

Fixes #436 #437 #438

Link to build : https://travis-ci.org/akrennmair/newsbeuter/builds/179715117

### CODE REVIEW

**Comment by tkerdonc** [comment]
Pushing only the test for now, to demonstrate they indeed fail. Will then push the fix.

File: src/rss.cpp [hunk 1]
```diff
@@ -302,17 +302,21 @@ void rss_item::set_flags(const std::string& ff) {
 void rss_item::sort_flags() {
 	std::sort(flags_.begin(), flags_.end());
 
-	for (auto it=flags_.begin(); flags_.size() > 0 && it!=flags_.end(); ++it) {
+	auto it=flags_.begin(); 
+	while (flags_.size() > 0 && it!=flags_.end()) {
 		if (!isalpha(*it)) {
 			flags_.erase(it);
 			it = flags_.begin();
```
Comment (Minoru) [inline src/rss.cpp]: Why rewind to the beginning when we can simply carry on as-is? (After `erase` `it` will point to the next character.)

File: src/rss.cpp [hunk 2]
```diff
@@ -302,17 +302,21 @@ void rss_item::set_flags(const std::string& ff) {
 void rss_item::sort_flags() {
 	std::sort(flags_.begin(), flags_.end());
 
-	for (auto it=flags_.begin(); flags_.size() > 0 && it!=flags_.end(); ++it) {
+	auto it=flags_.begin(); 
+	while (flags_.size() > 0 && it!=flags_.end()) {
```
Comment (Minoru) [inline src/rss.cpp]: The call to `size()` seems redundant to me.

File: src/rss.cpp [hunk 3]
```diff
@@ -302,17 +302,21 @@ void rss_item::set_flags(const std::string& ff) {
...
 		if (!isalpha(*it)) {
 			flags_.erase(it);
 			it = flags_.begin();
 		}
```
Comment (Minoru) [inline src/rss.cpp]: We put `else` on the same line as the closing bracket of the previous block: `} else {`.

File: src/rss.cpp [hunk 4]
```diff
@@ -302,17 +302,21 @@ void rss_item::set_flags(const std::string& ff) {
...
+		else {
+			it += 1;
+		}
 	}
 
 	for (unsigned int i=0; i<flags_.size(); ++i) {
 		if (i < (flags_.size()-1)) {
```
Comment (Minoru) [inline src/rss.cpp]: We could avoid that `if` entirely if we iterated while `i < flags_.size() - 1` (or so I believe; try and see what tests say :)

**Comment by Minoru** [comment]
Regarding tests: we create separate test files for separate source files. In this case, you're testing rss_item, which is defined in `include/rss.h`, so you should've created `test/rss.cpp` and put the tests there.

Regarding commits: I got your intention with demonstrating tests first, then adding a fix later, but I'd like to have a history where each commit does just one change. Currently your commit contains two separate changes; can you please split them?

**Comment by tkerdonc** [comment]
Thanks for the review. This latest push integrates the modification you requested, moves the test to rss.cpp and squashes the commits in two. Let me know what you think !

**Comment by tkerdonc** [comment]
Hmmm, tests don't look good, Should have tested the whole test suite before pushing. looking into that...

**Comment by tkerdonc** [comment]
Got it, the double comparison on line 316 is indeed required.
 ` 	for (unsigned int i=0; i<flags_.size(); ++i) { `
 Even if the test passes, we are accessing unallocated memory on the following line. Pushing the fix

File: src/rss.cpp [hunk 5]
```diff
@@ -302,17 +302,19 @@ void rss_item::set_flags(const std::string& ff) {
 void rss_item::sort_flags() {
 	std::sort(flags_.begin(), flags_.end());
 
-	for (auto it=flags_.begin(); flags_.size() > 0 && it!=flags_.end(); ++it) {
+	auto it=flags_.begin();
```
Comment (Minoru) [inline src/rss.cpp]: There's trailing whitespace on this line. Install [editorconfig](http://editorconfig.org/) plugin—it'll help with these details.

File: test/rss.cpp [hunk 6]
```diff
@@ -249,4 +249,23 @@ TEST_CASE("set_rssurl checks if query feed has a valid query", "[rss]") {
 	}
 }
 
+TEST_CASE("rss_item::sort_flags() cleans up flags", "[flag]") {
```
Comment (Minoru) [inline test/rss.cpp]: You forgot to update the tag (`"[flag]"`) here; it should say `"[rss]"`.

File: test/rss.cpp [hunk 7]
```diff
@@ -249,4 +249,23 @@ TEST_CASE("set_rssurl checks if query feed has a valid query", "[rss]") {
...
+	configcontainer cfg;
+	cache rsscache(":memory:", &cfg);
+	std::shared_ptr<rss_item> item = std::make_shared<rss_item>(&rsscache);
```
Comment (Minoru) [inline test/rss.cpp]: I just noticed: you don't really need a pointer here as you don't pass it to `rss_feed`'s methods.

**Comment by Minoru** [comment]
(Edited your comment to use [code block](https://help.github.com/articles/creating-and-highlighting-code-blocks/) instead of in-line code quotes.)

> the double comparison on line 316 is indeed required.

Yes, you're right; sorry for the silly suggestion.

Regarding commits: you updated the messages, but bundled all the changes into the second commit. Can you redo it so that the first commit contains first test and first fix, and the second commit contains second test and second fix?

**Comment by tkerdonc** [comment]
No worries, this was worth trying ^^. Splitted the commits properly this time, not sure what I was thinking writing messages not fitting the actual code changes, my bad.

File: test/rss.cpp [hunk 8]
```diff
@@ -249,4 +249,23 @@ TEST_CASE("set_rssurl checks if query feed has a valid query", "[rss]") {
 	}
 }
 
+TEST_CASE("rss_item::sort_flags() cleans up flags", "[rss]") {
+	configcontainer cfg;
+	cache rsscache(":memory:", &cfg);
+	rss_item item = rss_item(&rsscache);
```
Comment (Minoru) [inline test/rss.cpp]: Why not just `rss_item item(&rsscache)`?

Comment (tkerdonc) [inline test/rss.cpp]: Completely redundant indeed, this last push fixes it.

**Comment by Minoru** [comment]
Everything's in order; I'll merge once Travis checks finish (Travis has an outage at the moment, so this might take a while).

Thank you for working on this!

**Comment by tkerdonc** [comment]
You're welcome, thanks for the support and advice !

**Pull request merged**